### PR TITLE
Added message notifications when applying candidates.

### PR DESCRIPTION
**Pull request by dpomianbuff**
In this PR:
* Show messages in the UI when something goes wrong (i.e. no suggestion from ChatGPT, no applicable candidates, etc)
* Improve logging of candidates applicability
* Adjust code region selection bugfixes

### CODE REVIEW

File: src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt [hunk 1]
```diff
@@ -71,6 +76,7 @@ abstract class ApplyExtractFunctionTransformationIntention(
         logger.info("Invoking LLM with text: $text")
         val messageList = fewShotExtractSuggestion(text)
 
+        WindowManager.getInstance().getStatusBar(project).info = "Hello from Invoke LLM"
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt]: Is it visible for users? I suggest rephrasing the message to provide some useful details about what is happening right now :)

File: src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt [hunk 2]
```diff
@@ -80,15 +86,48 @@ abstract class ApplyExtractFunctionTransformationIntention(
...
-                            efCandidates.filter { isCandidateExtractable(it, editor, file) }.take(1).forEach { candidate ->
-                                codeTransformer.applyCandidate(candidate, project, editor, file)
+                        if (response.getSuggestions().isEmpty()) {
+                            showEFNotification(
+                                project,
+                                LLMBundle.message("notification.extract.function.with.llm.no.suggestions.message"),
+                                NotificationType.INFORMATION
+                            )
+                        } else {
+                            val efCandidateFactory = EFCandidateFactory()
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt]: Maybe we can extract code from this branch into a separate method?

File: src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt [hunk 3]
```diff
@@ -116,17 +160,26 @@ class EFCandidateFactory {
...
 
+        // if we are still on a PsiWhiteSpace, then go right
+        while (psiElement.getLineNumber(false) == psiElement.nextSibling?.getLineNumber(false) && psiElement is PsiWhiteSpace) {
+            psiElement = psiElement.nextSibling
+        }
+
         // if there are multiple parent PsiElements on the same line, look for the top one
-        while (psiElement.getLineNumber(false) == psiElement.parent?.getLineNumber(false)) {
-            psiElement = psiElement.parent
+        val ln = psiElement.getLineNumber(false)
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt]: Let's use the full names instead just `ln`

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt [hunk 4]
```diff
@@ -79,41 +87,107 @@ fun identifyExtractFunctionSuggestions(input: String): EFSuggestionList {
...
+    editor.selectionModel.setSelection(efCandidate.offsetStart, efCandidate.offsetEnd)
     var res = false
+    var reason = ""
+
+    if (selectionIsEntireBodyFunctionKotlin(efCandidate, editor, file)) {
+        observer?.update(
+            EFNotification(
+                result = EFApplicationResult.FAIL,
+                candidate = efCandidate,
+                reason = "Selection represents entire body function"
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: Should we move it to LLMBundle as well?

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt [hunk 5]
```diff
@@ -79,41 +87,107 @@ fun identifyExtractFunctionSuggestions(input: String): EFSuggestionList {
...
+        editor.selectionModel.removeSelection()
+        return false
+    }
+
+    if (elements.isEmpty()) {
+        observer?.update(
+            EFNotification(
+                result = EFApplicationResult.FAIL,
+                candidate = efCandidate,
+                reason = "Selected code can not be extracted in a new function"
```
Comment (onewhl) [inline src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: Should we move it to LLMBundle as well?

**Review by onewhl — CHANGES_REQUESTED** [review]
Cool, thank you Dorin! Please, take a look at the several comments I left

File: src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt (hunk 1 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt]: I have to remove this :)

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt (hunk 4 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: Yes. I'm going to move it to LLMBundle

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt (hunk 5 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: Yes. I'm going to move it to LLMBundle

File: src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt (hunk 2 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/intentions/ApplyExtractFunctionTransformationIntention.kt]: Done

File: src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt (hunk 3 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/EFCandidateFactory.kt]: Done

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt (hunk 4 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: Done

File: src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt (hunk 5 above)
Comment (dpomianbuff) [reply src/main/kotlin/com/intellij/ml/llm/template/utils/Utils.kt]: Done

**Pull request merged**